packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts

```diff
@@ -22,45 +23,156 @@ export class DataSyncMongoDBDataProvider<Type = any> extends MongoDBDataProvider
       throw e;
     });
     this.coerceTSFields = true;
+    this.conflictEngine = conflictEngine;
   }
 
   public async create(data: any, context: GraphbackContext): Promise<Type> {
     data._deleted = false;
+    if (this.conflictEngine !== undefined) {
+      const { next, conflictFieldName } = this.conflictEngine;
+      data[conflictFieldName] = next(undefined, data);
+    }
 
     return super.create(data, context);
   }
 
   public async update(data: any, context: GraphbackContext): Promise<Type> {
-    const conflict = await this.checkForConflicts(data, context);
 
-    if (conflict !== undefined) {
-      throw new ConflictError(conflict);
+    const { idField } = getDatabaseArguments(this.tableMap, data);
+
+    if (!idField.value) {
+      throw new NoDataError(`Cannot update ${this.collectionName} - missing ID field`)
     }
 
-    // TODO use findOneAndUpdate to check consistency afterwards
-    return super.update(data, context);
+    const objectId = new ObjectId(idField.value);
+
+    const startTime = Date.now();
+
+    const RETRY_LIMIT = 3000; // 3 seconds
+
+    while ((Date.now() - startTime) < RETRY_LIMIT) {   // Commented for debugging
+    // while (Date.now()) {
+
+      this.fieldTransformMap[TransformType.UPDATE]
+        .forEach((f: FieldTransform) => {
+          data[f.fieldName] = f.transform(data[f.fieldName]);
+        });
+
+      const updateFilter: any = { _id: objectId, _deleted: { $ne: true } };
+      const updateSets = Object.assign({}, data);
+      // eslint-disable-next-line @typescript-eslint/tslint/config
+      delete updateSets[this.tableMap.idField];
+
+      if (this.conflictEngine !== undefined) {
+        
+        const { conflictFieldName, next } = this.conflictEngine;
+        const currentState = updateSets[conflictFieldName];
+        updateFilter[conflictFieldName] = currentState;
+
+        // eslint-disable-next-line @typescript-eslint/tslint/config
+        delete updateSets[conflictFieldName]
+
+        updateSets[conflictFieldName] = next(currentState, updateSets);
+      }
+
+      const updateResult = await this.db.collection(this.collectionName).findOneAndUpdate(updateFilter, {
+        $set: updateSets
+      },{
+        w: "majority",
+        returnOriginal: false
+      })
+
+      if (!updateResult.value) {
+        // There is conflict or document doesn't exist
+  
+        // Check if document exists
```

Comment:
This call makes no sene for me. We checking if document exist after we checked for conflict?

Can you put some text how conflict is being checked.

Reply:
if the above findOneAndUpdate fails, updateResult.value will be `null`, that means that there was no update done,
this means that either the document does not exist in the collection or the version did not match,
next we check if the document exists in the collection using findOne, if the document exists, then the reason update failed is because version did not match

Reply:
This contradics what we shown in initial documentation that we can use upserts. 
So this approach was applied without much comments why upsert did not work etc. 

Extra queries bring performance impact, so if you write in meta language what is suggested way to handle update/delete we can consult that with mongodb team and make sure that we getting the most performant option rather than something that just works.

Reply:
Apologies, this was a shortcoming I forgot to discuss, this falls on me. My problem with using upserts: If we send an updateComment mutation with invalid id like anything garbage, upsert:true would create that document and we would likely end up with garbage in the database. I would certainly like to discuss about current approach and other possible approaches.